### ISSUE TITLE
[FEATURE]Add `dedup` PPL Command

### ISSUE THREAD

**Issue by YANG-DB** [opensearch-project/opensearch-spark#421]
**Is your feature request related to a problem?**
Add `dedup` command for PPL spark based execution driver

[PPL Dedup](https://github.com/opensearch-project/sql/blob/d3ecb2d75712db2bcbd2945e62728c6599580657/docs/user/ppl/cmd/dedup.rst)

**Do you have any additional context?**
 - https://github.com/opensearch-project/opensearch-spark/issues/408
 - https://github.com/opensearch-project/sql/blob/d3ecb2d75712db2bcbd2945e62728c6599580657/docs/user/ppl/cmd/dedup.rst

**Comment by LantaoJin** [comment]
I'm starting to contribute it.

**Comment by LantaoJin** [comment]
## Design Doc

__Syntax__
```
dedup [int] <field-list> keepempty=<bool>] [consecutive=<bool>]
```
`int` also known as `allowedDuplication`, the number of duplicate events must be greater than 0.

We can separate this feature to 2 parts:
- `allowedDuplication = 1`
- `allowedDuplication > 1`

### Part 1

For `allowedDuplication = 1`,  we translate following three PPL `dedup` commands to different Spark Logical Plans. (assuming the \<field-list\> is `a, b`)
1. keepempty=false: `| dedup [1] a, b [keepempty=false]`
```
Deduplicate ['a, 'b]
+- Filter (isnotnull('a) AND isnotnull('b))
     +- Project
          +- UnresolvedRelation
```
2. keepempty=true : `| dedup [1] a, b keepempty=true`
```
Union
:- Deduplicate ['a, 'b]
:   +- Filter (isnotnull('a) AND isnotnull('b))
:        +- Project
:             +- UnresolvedRelation
+- Filter (isnull('a) OR isnull('b))
     +- Project
          +- UnresolvedRelation
```

3. consecutive=true: `| dedup [1] a, b [keepempty=true] consecutive=true`
```
UnsupportedOperationException("Consecutive deduplication is not supported")
```

#### Question
__`Deduplicate` or `DeduplicateWithinWatermark`?__

https://issues.apache.org/jira/browse/SPARK-42931 introduced a new LogicalPlan `DeduplicateWithinWatermark` to deduplicate events within the watermark and was available in 3.5.0+
I am not clear the architecture of Flint, but seems it performs on Spark Streaming. Deduplication on streaming dataset has different semantic with batch dataset:
> We document the behavior clearly that the event time column should be a part of the subset columns for deduplication to clean up the state, but it cannot be applied to the customers as timestamps are not exactly the same for duplicated events in their use cases.

Need more guidance on the purpose of this repository and how it should be used in production or in future. For current time-being, I choose `Deduplicate` in this PR (`DeduplicateWithinWatermark` required 3.5.0+).

### Part 2

To translate `dedup` command with `allowedDuplication > 1`, such as `| dedup 2 a,b` to Spark plan, the solution is translating to a plan with Window function (e.g row_number) and a new column `row_number_col` as Filter.
- For `| dedup 2 a, b keepempty=false`
```
Project ['a, 'b] // ensure the row_number_col is removed. not a and b only, should contain all fields except row_number_col
+- Filter ('row_number_col <= 2) // allowed duplication = 2
   +- Window [row_number() windowspecdefinition('a, 'b, 'order_key ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS 'row_number_col], ['a, 'b], ['order_key ASC NULLS FIRST]
       +- Filter (isnotnull('a) AND isnotnull('b)) // keepempty=false
          +- Project
             +- UnresolvedRelation
```
- For `| dedup 2 a, b keepempty=true`
```
Union
:- Project ['a, 'b]
:  +- Filter ('row_number_col <= 2)
:     +- Window [row_number() windowspecdefinition('a, 'b, 'order_key ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS 'row_number_col], ['a, 'b], ['order_key ASC NULLS FIRST]
:        +- Filter (isnotnull('a) AND isnotnull('b))
:           +- Project
:              +- UnresolvedRelation
+- Filter (isnull('a) OR isnull('b))
   +- Project
      +- UnresolvedRelation
```

#### Question

__What is the `'order_key` in Window node?__
AFAIK, there always should be a time series field existing for log analysis, such as `_time` or `@timestamp`. How to get it in AST?
If we cannot point out an existing time series field as `order_key`, the syntax of `dedup` command might change to
```
| dedup 2 a, b order by c
| dedup 2 a, b orderkey=c
```

FYI: In Spark, window function requires window to be ordered. Query fails if order by clause is not specified. For example, `SELECT <wf>(expr) OVER ([PARTITION BY window_partition] ORDER BY window_ordering) from table`
But in MySQL or opensearch-sql, the order by clause can be omitted.
